contracts/tec/contracts/src/MixinTECApprovalVerifier.sol

```diff
@@ -0,0 +1,200 @@
+                continue;
+            }
+            
+            // Ethereum transaction signer gives implicit signature of approval
+            address approverAddress = orders[i].feeRecipientAddress;
+            if (approverAddress == tx.origin) {
+                approvalSignerAddresses = approvalSignerAddresses.append(tx.origin);
+                continue;
+            }
+
+            // Ensure feeRecipient of order has approved this 0x transaction
+            bool isOrderApproved = approvalSignerAddresses.contains(approverAddress);
+            require(
+                isOrderApproved,
+                "INVALID_APPROVAL_SIGNATURE"
+            );
+
+            // The Ethereum transaction signer must be the 0x transaction signer or an approver of the 0x transaction
+            require(
+                transaction.signerAddress == tx.origin || approvalSignerAddresses.contains(tx.origin),
+                "INVALID_SENDER"
+            );
+        }
+    }
+
+    /// @dev Decodes the orders from Exchange calldata representing any fill method.
+    /// @param data Exchange calldata representing a fill method.
+    /// @return The orders from the Exchange calldata.
+    function decodeFillDataOrders(bytes memory data)
+        internal
+        pure
+        returns (LibOrder.Order[] memory orders)
+    {
+        bytes4 selector = data.readBytes4(0);
+        if (
+            selector == FILL_ORDER_SELECTOR ||
+            selector == FILL_ORDER_NO_THROW_SELECTOR ||
+            selector == FILL_OR_KILL_ORDER_SELECTOR
+        ) {
+            // Decode single order
+            (LibOrder.Order memory order) = abi.decode(
+                data.slice(4, data.length),
+                (LibOrder.Order)
+            );
+            orders = new LibOrder.Order[](1);
+            orders[0] = order;
+        } else if (
+            selector == BATCH_FILL_ORDERS_SELECTOR ||
+            selector == BATCH_FILL_ORDERS_NO_THROW_SELECTOR ||
+            selector == BATCH_FILL_OR_KILL_ORDERS_SELECTOR ||
+            selector == MARKET_BUY_ORDERS_SELECTOR ||
+            selector == MARKET_BUY_ORDERS_NO_THROW_SELECTOR ||
+            selector == MARKET_SELL_ORDERS_SELECTOR ||
+            selector == MARKET_SELL_ORDERS_NO_THROW_SELECTOR
+        ) {
+            // Decode all orders
+            // solhint-disable indent
+            (orders) = abi.decode(
+                data.slice(4, data.length),
+                (LibOrder.Order[])
+            );
+        } else if (selector == MATCH_ORDERS_SELECTOR) {
+            // Decode left and right orders
+            (LibOrder.Order memory leftOrder, LibOrder.Order memory rightOrder) = abi.decode(
+                data.slice(4, data.length),
+                (LibOrder.Order, LibOrder.Order)
+            );
+
+            // Create array of orders
+            orders = new LibOrder.Order[](2);
+            orders[0] = leftOrder;
+            orders[1] = rightOrder;
+        }
```

Comment:
Should we handle the case of an unrecognized selector? Handling this case explicitly would future-proof this code, by ensuring it can’t call functions that it doesn’t know how to handle.

Reply:
As is, an unknown selector will return an empty `Order` array. If that isn't a valid Exchange selector, then the call will fail when it hits the Exchange. I think this is the simplest way of allowing all non-fill functions to be automatically approved, rather than hard coding in the allowed functions.

The default behavior of this contract is to not duplicate any validation that already occurs in the Exchange contract in order to reduce cost/complexity.

Reply:
Yeah so the scenario I was thinking of is - suppose this contract were used with a future version of the exchange, which includes a new set of functions that must be validated by a TEC. These new functions would be automatically approved, when they should be validated. 

While we specify the exchange in the constructor, there's no hard link between this version of the TEC and a compatible version of the exchange.

One way to mitigate this is to have a whitelist of known functions ([this](https://github.com/0xProject/0x-monorepo/blob/development/contracts/extensions/contracts/src/BalanceThresholdFilter/MixinBalanceThresholdFilterCore.sol#L94) list should suffice). Alternatively, we could include a call in the constructor to assert that `Exchange.VERSION` is `2.0.0`, although this would be less robust.